File: cpp/202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        int slow = n;
        int fast = getNext(n);

        while (slow != fast and fast != 1) {
            slow = getNext(slow);
            fast = getNext(getNext(fast));
        }

        if (fast == 1) {
            return true;
        }        
        return false;
    }
private:
    int getNext(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            n = n / 10;
            sum += digit * digit;
        }
        return sum;
    }
};

File: rust/202-happy-number.rs
impl Solution {
    pub fn is_happy(mut n: i32) -> bool {
        loop {
            let mut sum = 0;
            while n > 0 {
                sum += (n % 10).pow(2);
                n = n / 10;
            }
            match sum {
                1 | 4 => break sum == 1,
                _ => n = sum,
            }
        }
    }
}